File: agents/python/shared/guardrails/sanitize.py
"""净化不可信文本中的存储型和间接提示注入。

评论、描述或订单备注可能包含忽略指令、角色伪造、虚假系统轮次或
窃取提示词的内容。用 [neutralized] 替换相关标记，保留文本曾存在的
证据，并移除用于隐藏指令的控制字符和零宽字符。

纯函数，无 I/O、模型或数据库访问；规则偏重低误报，与提示词规则
及输入注入检测组成多层防护。
"""

from __future__ import annotations

import re
from typing import Any

# 移除 C0 控制符，但保留 TAB、LF 和 CR；
# 另移除 DEL、零宽字符、行段分隔符及 BOM。
# 通过数字码点和 translate 声明，
# 避免源码直接包含控制字节。
_STRIP_CODEPOINTS = (
    *range(0x00, 0x09),
    0x0B,
    0x0C,
    *range(0x0E, 0x20),
    0x7F,
    *range(0x200B, 0x2010),
    0x2028,
    0x2029,
    0xFEFF,
)
_CONTROL_TRANSLATION = dict.fromkeys(_STRIP_CODEPOINTS)

# 高精度注入信号，不区分大小写；伪造轮次规则锚定行首，
# 避免误伤普通正文，
# 例如商品名称中的 System: Pro。
_INJECTION_PATTERNS: tuple[re.Pattern[str], ...] = (
    re.compile(
        r"ignore\s+(?:all\s+|any\s+)?(?:the\s+)?(?:previous|prior|above|earlier)\s+"
        r"(?:instructions?|prompts?|rules?|messages?)",
        re.I,
    ),
    re.compile(
        r"disregard\s+(?:all\s+|any\s+)?(?:the\s+)?(?:previous|prior|above|earlier)\s+"
        r"(?:instructions?|prompts?|rules?)",
        re.I,
    ),
    re.compile(
        r"forget\s+(?:all\s+|everything\s+|your\s+)?(?:previous\s+|prior\s+)?"
        r"(?:instructions?|rules?)",
        re.I,
    ),
    re.compile(r"you\s+are\s+now\s+(?:a|an|the)\b", re.I),
    re.compile(r"new\s+(?:system\s+)?(?:instructions?|prompts?|rules?)\s*:", re.I),
    re.compile(r"(?m)^\s{0,6}(?:system|developer|assistant)\s*:", re.I),
    re.compile(r"reveal\s+(?:your\s+|the\s+)?(?:system\s+)?(?:prompt|instructions?)", re.I),
    re.compile(r"</?\s*(?:system|instructions?|prompt)\s*>", re.I),
    re.compile(r"\bact\s+as\s+(?:if\s+you\s+are\s+)?(?:an?\s+)?admin", re.I),
)

_MARK = "[neutralized]"


def contains_injection_markers(text: str) -> bool:
    """文本命中任一高精度注入信号时返回 True。"""
    if not text:
        return False
    return any(pattern.search(text) for pattern in _INJECTION_PATTERNS)


def neutralize_text(text: str) -> str:
    """净化单个字符串的注入标记，并移除控制与零宽字符。"""
    if not text:
        return text
    cleaned = text.translate(_CONTROL_TRANSLATION)
    for pattern in _INJECTION_PATTERNS:
        cleaned = pattern.sub(_MARK, cleaned)
    return cleaned


def neutralize_value(value: Any, *, fields: set[str] | None = None, _key: str | None = None) -> Any:
    """递归净化工具结果中的不可信字符串。

    value 为字符串、字典、列表或标量。fields 指定时，只净化任意深度
    直属字典键命中的字符串；未指定则处理全部字符串。_key 为递归内部
    使用的当前字段名。
    """
    if isinstance(value, str):
        if fields is None or _key in fields:
            return neutralize_text(value)
        return value
    if isinstance(value, dict):
        return {k: neutralize_value(v, fields=fields, _key=k) for k, v in value.items()}
    if isinstance(value, list):
        return [neutralize_value(v, fields=fields, _key=_key) for v in value]
    if isinstance(value, tuple):
        return tuple(neutralize_value(v, fields=fields, _key=_key) for v in value)
    return value

File: agents/dotnet/src/ECommerceAgents.Shared/Guardrails/Sanitize.cs
using System.Text.RegularExpressions;

namespace ECommerceAgents.Shared.Guardrails;

/// <summary>
/// Neutralize stored / indirect prompt injection in untrusted text — the
/// .NET twin of Python's <c>shared/guardrails/sanitize.py</c>.
/// </summary>
/// <remarks>
/// Tool results re-enter the model as function-result content. Content that
/// originates from other users — product reviews, descriptions, order notes
/// — can carry adversarial instructions ("ignore previous instructions",
/// role reassignments, fake system turns, system-prompt exfiltration). This
/// class defangs those markers (replaces them with an inert
/// <c>[neutralized]</c> token rather than deleting them, so legitimate
/// analysis still sees the text existed).
///
/// Pure functions: no I/O, no LLM, no DB. Patterns are deliberately high
/// precision (low false-positive) — the prompt-layer rules are the other
/// layer of defense. Patterns copied verbatim from the Python source so the
/// two stacks flag the same inputs.
/// </remarks>
public static class Sanitize
{
    private const string Mark = "[neutralized]";

    // Codepoints to strip: C0 controls except TAB (0x09) / LF (0x0A) / CR (0x0D),
    // DEL, the zero-width marks, line/paragraph separators, and the BOM —
    // used to smuggle hidden instructions past a naive text scan. Same set as
    // Python's sanitize.py, declared numerically for the same reason: the
    // source file never contains raw control bytes.
    private static readonly HashSet<int> StripCodepoints = BuildStripCodepoints();

    private static HashSet<int> BuildStripCodepoints()
    {
        var set = new HashSet<int>();
        for (var c = 0x00; c < 0x09; c++) set.Add(c);
        set.Add(0x0B);
        set.Add(0x0C);
        for (var c = 0x0E; c < 0x20; c++) set.Add(c);
        set.Add(0x7F);
        for (var c = 0x200B; c < 0x2010; c++) set.Add(c);
        set.Add(0x2028);
        set.Add(0x2029);
        set.Add(0xFEFF);
        return set;
    }

    private static readonly Regex[] InjectionPatterns =
    [
        new(
            @"ignore\s+(?:all\s+|any\s+)?(?:the\s+)?(?:previous|prior|above|earlier)\s+" +
            @"(?:instructions?|prompts?|rules?|messages?)",
            RegexOptions.IgnoreCase | RegexOptions.Compiled
        ),
        new(
            @"disregard\s+(?:all\s+|any\s+)?(?:the\s+)?(?:previous|prior|above|earlier)\s+" +
            @"(?:instructions?|prompts?|rules?)",
            RegexOptions.IgnoreCase | RegexOptions.Compiled
        ),
        new(
            @"forget\s+(?:all\s+|everything\s+|your\s+)?(?:previous\s+|prior\s+)?" +
            @"(?:instructions?|rules?)",
            RegexOptions.IgnoreCase | RegexOptions.Compiled
        ),
        new(@"you\s+are\s+now\s+(?:a|an|the)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"new\s+(?:system\s+)?(?:instructions?|prompts?|rules?)\s*:", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"^\s{0,6}(?:system|developer|assistant)\s*:", RegexOptions.IgnoreCase | RegexOptions.Compiled | RegexOptions.Multiline),
        new(@"reveal\s+(?:your\s+|the\s+)?(?:system\s+)?(?:prompt|instructions?)", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"</?\s*(?:system|instructions?|prompt)\s*>", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        new(@"\bact\s+as\s+(?:if\s+you\s+are\s+)?(?:an?\s+)?admin", RegexOptions.IgnoreCase | RegexOptions.Compiled),
    ];

    /// <summary>Returns true if <paramref name="text"/> matches any high-precision injection signal.</summary>
    public static bool ContainsInjectionMarkers(string? text) =>
        !string.IsNullOrEmpty(text) && InjectionPatterns.Any(p => p.IsMatch(text));

    /// <summary>Strips smuggled control/zero-width characters and defangs injection markers.</summary>
    public static string NeutralizeText(string? text)
    {
        if (string.IsNullOrEmpty(text))
        {
            return text ?? string.Empty;
        }

        var stripped = new System.Text.StringBuilder(text.Length);
        foreach (var ch in text)
        {
            if (!StripCodepoints.Contains(ch))
            {
                stripped.Append(ch);
            }
        }

        var cleaned = stripped.ToString();
        foreach (var pattern in InjectionPatterns)
        {
            cleaned = pattern.Replace(cleaned, Mark);
        }
        return cleaned;
    }
}
